src/main/java/com/wultra/security/userdatastore/service/AttachmentService.java

```diff
@@ -0,0 +1,143 @@
+import com.wultra.security.userdatastore.model.response.AttachmentCreateResponse;
+import com.wultra.security.userdatastore.model.response.AttachmentResponse;
+import com.wultra.security.userdatastore.model.response.PhotoCreateResponse;
+import com.wultra.security.userdatastore.model.response.PhotoResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Service for document attachments.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
+@Slf4j
+public class AttachmentService {
+
+    private final AttachmentRepository attachmentRepository;
+    private final DocumentRepository documentRepository;
+    private final Audit audit;
+    private final EncryptionService encryptionService;
+    private final AttachmentConverter attachmentConverter;
+
+    @Autowired
+    AttachmentService(final AttachmentRepository attachmentRepository, final DocumentRepository documentRepository, final Audit audit, final EncryptionService encryptionService, final AttachmentConverter attachmentConverter) {
+        this.attachmentRepository = attachmentRepository;
+        this.documentRepository = documentRepository;
+        this.audit = audit;
+        this.encryptionService = encryptionService;
+        this.attachmentConverter = attachmentConverter;
+    }
+
+    @Transactional
+    public AttachmentCreateResponse createAttachment(final String userId, final String documentId, final AttachmentCreateRequest request) {
+        final Optional<DocumentEntity> documentEntityOptional = documentRepository.findById(documentId);
+        if (documentEntityOptional.isEmpty()) {
+            throw new ResourceNotFoundException("Document not found, ID: '%s'".formatted(documentId));
+        }
+        final DocumentEntity documentEntity = documentEntityOptional.get();
+        final AttachmentEntity attachmentEntity = new AttachmentEntity();
+        attachmentEntity.setId(UUID.randomUUID().toString());
+        attachmentEntity.setDocument(documentEntity);
+        attachmentEntity.setAttachmentType(request.attachmentType());
+        attachmentEntity.setAttachmentData(request.attachmentData());
+        attachmentEntity.setTimestampCreated(LocalDateTime.now());
+
+        attachmentRepository.save(attachmentEntity);
+
+        final AttachmentCreateResponse response = new AttachmentCreateResponse();
+        response.setId(attachmentEntity.getId());
+        response.setDocumentId(documentEntity.getId());
+        return response;
+    }
+
+    public List<AttachmentDto> fetchAttachments(final String userId, final String documentId) {
```

Comment:
What about adding `@Transactional(readOnly = true)`?

Reply:
Thanks, yes the annotation was missing.